File: ruby/say/say.rb
class Say
  def initialize(n)
    @n = n
  end

  def in_english
    raise ArgumentError, "Invalid number" unless (0...1_000_000_000_000) === @n

    return "zero" if @n.zero?

    r = @n
    chunks = []
    while r != 0
      r, n = r.divmod 1_000
      chunks << less_than_one_thousand(n)
    end
    chunks.zip(SCALES).filter { |n, *| !n.empty? }.reverse.flatten.compact.join(THOUSAND_SEPARATOR)
  end

  private

  SCALES = [nil, "thousand", "million", "billion"]
  HUNDRED = "hundred"
  TENS = %w[twenty thirty forty fifty sixty seventy eighty ninety]
  ONES = %w[one two three four five six seven eight nine ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen]
  THOUSAND_SEPARATOR = " "
  HUNDRED_SEPARATOR = " "
  TEN_SEPARATOR = "-"

  def less_than_one_thousand(n)
    h, l = n.divmod 100
    r = []
    r << less_than_one_hundred(h) << HUNDRED if h.nonzero?
    r << less_than_one_hundred(l) if l.nonzero?
    r.join(HUNDRED_SEPARATOR)
  end

  def less_than_one_hundred(n)
    return less_than_twenty(n) if n < 20
    t, l = n.divmod 10
    r = [TENS[t - 2]]
    r << less_than_twenty(l) if l.nonzero?
    r.join(TEN_SEPARATOR)
  end

  def less_than_twenty(n)
    ONES[n - 1]
  end
end

File: clojure/say/src/say.clj
(ns say
  (:require [clojure.string :as string]))

(def ^:private ones
  ["zero" "one" "two" "three" "four" "five" "six" "seven" "eight", "nine", "ten"
   "eleven" "twelve" "thirteen" "fourteen" "fifteen" "sixteen" "seventeen" "eighteen" "nineteen"])
(def ^:private tens
  ["twenty" "thirty" "forty" "fifty" "sixty" "seventy" "eighty" "ninety"])
(def ^:private scales
  [nil "thousand" "million" "billion"])

(defn- divmod [num div]
  ((juxt quot rem) num div))

(defn- conj-if [coll test x & xs]
  (if test (apply conj coll x xs) coll))

(defn- less-than-20 [num]
  (get ones num))

(defn- less-than-100 [num]
  (if (< num 20)
    (less-than-20 num)
    (let [[t l] (divmod num 10)]
      (as-> [(get tens (- t 2))] $
            (conj-if $ (pos? l) (less-than-20 l))
            (string/join "-" $)))))

(defn- less-than-1000 [num]
  (let [[h l] (divmod num 100)]
    (as-> [] $
          (conj-if $ (pos? h) (less-than-100 h) "hundred")
          (conj-if $ (pos? l) (less-than-100 l))
          (string/join " " $))))

(defn- parts [num]
  (loop [r num
         p []]
    (if (zero? r)
      p
      (let [[r n] (divmod r 1000)]
        (recur r (conj p (less-than-1000 n)))))))

(defn number [num]
  (when-not (< -1 num 1000000000000)
    (throw (IllegalArgumentException. "num must be between 0 and 999999999999")))
  (if (zero? num)
    (get ones 0)
    (->> (parts num)
         (#(map vector % scales))
         (filter #(not (string/blank? (first %))))
         reverse
         flatten
         (filter some?)
         (string/join " "))))
